### PR TITLE
add debugging support to Python language host

### PR DESCRIPTION
**Pull request by tgummerer**
Add debugging support for the Python language. The scaffolding for this was put in place previously in #17072 and then #17087.

To do this we start the program under debugpy, and then wait for the debugger to get ready before informing the engine that it needs to inform the user that a debugger should be attached.

Fixes https://github.com/pulumi/pulumi/issues/17172

### CODE REVIEW

File: sdk/python/toolchain/pip.go [hunk 1]
```diff
@@ -446,11 +446,12 @@ func InstallDependencies(ctx context.Context, cwd, venvDir string, useLanguageVe
...
 
-	printmsg("Updating pip, setuptools, and wheel in virtual environment...")
+	printmsg("Updating pip, setuptools, wheel, and debugpy in virtual environment...")
 
 	// activate virtual environment
-
-	err := runPipInstall("updating pip, setuptools, and wheel", "--upgrade", "pip", "setuptools", "wheel")
+	err := runPipInstall(
+		"updating pip, setuptools, wheel, and debugpy",
+		"--upgrade", "pip", "setuptools", "wheel", "debugpy")
```
Comment (julienp) [inline sdk/python/toolchain/pip.go:454]: When using Poetry we don't run this code. We don't need these when **using** a pulumi SDK, we only need them for packaging & publishing SDKs. I don't think we should have this code at all, it should probably be added to requirements.txt by the SDK codegen, but it is here now.

To help with the Poetry case, maybe when starting in debug mode, we can first call `toolchain.ListPackages` and check if debugpy is installed. If it is not present, we can tell the user to first install it (`pip install debugpy` or `poetry add debugpy`) and exit with an error.

If we do always want `debugpy` to be available, then we should add it as a dependency of `pulumi` itself.

**Review by julienp — APPROVED** [review]

File: sdk/python/toolchain/pip.go (hunk 1 above)
Comment (tgummerer) [reply sdk/python/toolchain/pip.go:454]: Ah thanks for that!  Yeah `toolchain.ListPackages` sounds like a great idea, and then we can let the user deal with it.  That's probably better than forcing every user to install `debugpy` even if they might not need it.

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 2]
```diff
@@ -676,14 +693,185 @@ func determinePluginVersion(packageVersion string) (string, error) {
...
+type NotInstalledError struct {
+	InstallMessage string
+}
+
+func (e *NotInstalledError) Error() string {
+	return e.InstallMessage
+}
+
+func checkForPackage(ctx context.Context, pkg string, opts toolchain.PythonOptions) error {
+	// If the typechecker is not installed, tell the user to install it.
```
Comment (julienp) [inline sdk/python/cmd/pulumi-language-python/main.go:819]:
```suggestion
	// If the typechecker is not installed, tell the user to install it.
```

**Review by julienp — APPROVED** [review]
Nice!

File: sdk/python/cmd/pulumi-language-python/main.go (hunk 2 above)
Comment (tgummerer) [reply sdk/python/cmd/pulumi-language-python/main.go]: Whoops, thanks for catching that!

**Pull request merged**